### DISCUSSION TITLE
LaTeX Macros

### DISCUSSION THREAD

**Discussion by claudiushaag** [marimo-team/marimo#1941] [Ideas]
Hi to all!

Besides presenting results and inspecting data, we're using notebooks for teaching in our research group.
As we're growing very fond of marimo and are enthusiastic about all its possiblities and features,  we would like to also switch our notebook-usage for teaching from jupyter to marimo-notebooks.

But our teaching involves a lot of formulas and special characters, which we define as LaTeX-macros. In the past, we have found a way of defining those for MathJax (using them in JupyterLab), and recently, we also found a way of defining macros for KaTeX, so we're able to display our Math in the VSCode Notebook-Editor. Our solution looks like this:

```
import IPython.display
IPython.display.display_latex(IPython.display.Latex(filename="DAE_katex_macros.tex"))
```
with `DAE_katex.macros.tex`  looking like:
```
\let\DAEnewcommand\outer\renewcommand{\DAEnewcommand}[2]{\let#1\outer\renewcommand{#1}{#2}}
\DAEnewcommand{\invp}{{\it Inv}^{\small +}}
```

This enables us to use our KaTeX-macros in the Jupyter Markdown Cells, for example like this:
```
$\invp$
```
Sadly, this workaround does not work in a marimo notebook, but we think this would be a valuable feature, as we love the app view and think it could really change the effectiveness of our teaching. In regards to science-communication, effective ways to present formulas might also come in handy. On a sidenote: We think that MathJax provides the larger feature-set compared to KaTeX.

Thank you so much for an amazing tool!

**Comment by akshayka** [comment]
Thanks for the kind words!

Totally understand where you're coming from. My lab had an extensive macros file. Some thoughts.

- For my own reference (and for anyone else who looks at this), here's how macros work in IPython/Jupyter: https://groups.google.com/g/jupyter/c/Wlu-eS5LlTo/m/cTnwoq2eBAAJ
- I don't think we can use the same mechanism, or at least I personally don't see how, perhaps someone else sees it. We use KaTeX programmatically, maybe the Jupyter implementation uses KaTeX's autorender (https://katex.org/docs/autorender).
- KaTeX can be configured with macros (https://katex.org/docs/options). Perhaps we could expose that.

**Reply by claudiushaag** [reply]
Thank you for your fast reply and acknowledgement of our sitation!

I fiddled a little bit more with trying to use macros: Currently, Marimo does not even provide cell-persistent KaTeX macros.
I'm not able to define one and use it in a different cell. Otherwise, I could have envisioned some work-around by reading in and defining the macros in a f-string in a markdown-cell.

It seems like ``globalGroup`` is the relevant setting in the ``katex.render()``-call, but I'm completely unaware of its side-effects and security-implications.

**Comment by claudiushaag** [comment]
@akshayka Love the way marimo has grown in both features and users! What is your impression on the potential timeline of this feature?
This is something that still witholds our transition from jupyter for a lot of stuff. It also render's a few really cool features as serving the notebooks on github unactractive for us, as there is no feasible way of explaining our work.
I do not have any experience developing typescript, so I'm sadly not able to contribute without a few more pointers.
